Libraries/LibWeb/Credentials/CredentialsContainer.cpp

```diff
@@ -0,0 +1,127 @@
+    if (!document.is_fully_active()) {
+        return create_rejected_promise_from_exception(realm, WebIDL::InvalidStateError::create(realm, "Document is not fully active yet"_string));
+    }
+
+    // 5. If options.signal is aborted, then return a promise rejected with options.signal’s abort reason.
+    if (options.signal && options.signal->aborted()) {
+        return WebIDL::create_rejected_promise(realm, options.signal->reason());
+    }
+
+    // 6. Let interfaces be option’s relevant credential interface objects.
+    auto const interfaces = relevant_credential_interface_objects(options);
+
+    // 7. If interfaces is empty, then return a promise rejected with a "NotSupportedError" DOMException.
+    if (interfaces.is_empty()) {
+        return create_rejected_promise_from_exception(realm, WebIDL::DOMException::create(realm, "NotSupportedError"_string, "No credential type is supported"_string));
+    }
+
+    // FIXME
+    // 8. For each interface of interfaces:
+    {
+        // FIXME
+        // 1. If options.mediation is conditional and interface does not support conditional user mediation, return a promise rejected with a "TypeError" DOMException.
+        if (options.mediation == Bindings::CredentialMediationRequirement::Conditional) {
+            return create_rejected_promise_from_exception(realm, WebIDL::DOMException::create(realm, "TypeError"_string, "Interface does not support user mediation"_string));
+        }
+
+        // FIXME
+        // 2. If settings’ active credential types contains interface’s [[type]], return a promise rejected with a "NotAllowedError" DOMException.
+        // 3. Append interface’s [[type]] to settings’ active credential types.
+    }
+
+    // 9. Let origin be settings’ origin.
+    auto const& origin = settings.origin();
+
+    // FIXME
+    // 10. Let sameOriginWithAncestors be true if settings is same-origin with its ancestors, and false otherwise.
+
+    // FIXME
+    // 11. For each interface in options’ relevant credential interface objects:
+    {
+        // 1. Let permission be the interface’s [[type]] Get Permissions Policy.
+        // 2. If permission is null, continue.
+        // 3. If document is not allowed to use permission, return a promise rejected with a "NotAllowedError" DOMException.
+    }
+
+    // 12. Let p be a new promise.
+    auto p = WebIDL::create_promise(realm);
+
+    // FIXME
+    // 13. Run the following steps in parallel:
+
+    // FIXME
+    // 14. React to p
+
+    // 15. Return p
+    return p;
+}
+
+CredentialsContainer::CredentialsContainer(JS::Realm& realm)
+    : PlatformObject(realm)
+{
+}
+
+void CredentialsContainer::initialize(JS::Realm& realm)
+{
+    Base::initialize(realm);
+    WEB_SET_PROTOTYPE_FOR_INTERFACE(CredentialsContainer);
+}
+
+// https://www.w3.org/TR/credential-management-1/#credentialrequestoptions-dictionary
+HashTable<Credential> relevant_credential_interface_objects(Variant<CredentialCreationOptions, CredentialRequestOptions>)
```

Comment:
Technically it is an error to have a value of a GC::Cell stored like this. It should always be wrapped in a GC::Ptr, GC::Ref, or (if really necessary to interact with the non-gc world) GC::Handle.

This method feels like the spec's idea and our idea of what is going on will differ.

For an example of how we've handled this kind of hand waving around interfaces before, you might look at the WebCrypto stuff, especially with regards to the Algorithms. IndexedDB does this kind of thing too.

Either way, we'll come across that issue when this is implemented fully, so I'm not going to sweat it too much